### PR TITLE
Cleanup of GoogleAuthenticator

### PR DESCRIPTION
**Pull request by iltar**
This PR fixes the following things:
 - `hash_equals` as build-in php function to compare strings instead of a custom variant
 - A clearer `getUrl` method in preparation of a change require to make it more compatible with the google authenticator
 - php 7 notation shortcuts
 - php 7 return types
 - A new (optional) constructor argument indicating now, to prevent flaky tests when testing with codes
 - Fixed a broken unit-test/bug where false was considered as a non-argument for `getCode`
 - Cleaned up a private property that couldn't have possibly be used anywhere
 - Added more tests to prove the boundries of `checkCode`
 - Removed an unused variable from the example

In a followup PR I will work on making the actual functionality closer to what the google authenticator provides.

### CODE REVIEW

File: src/GoogleAuthenticator.php [hunk 1]
```diff
@@ -143,36 +148,6 @@ public function generateSecret()
...
-     * A constant time code comparison.
-     *
-     * @param string $known known code
-     * @param string $given code received from a user
-     *
-     * @return bool
-     *
-     * @see http://codereview.stackexchange.com/q/13512/6747
-     */
-    private function codesEqual(string $known, string $given): bool
```
Comment (greg0ire) [inline src/GoogleAuthenticator.php]: Should be deprecated, not removed.

Comment (greg0ire) [inline src/GoogleAuthenticator.php]: Oh nvm it's private

Comment (iltar) [inline src/GoogleAuthenticator.php]: It's a private function that's only used at 1 location

**Comment by greg0ire** [comment]
Tagging as minor because of the signature improvement

File: src/GoogleAuthenticator.php [hunk 2]
```diff
@@ -113,26 +117,27 @@ public function getCode($secret, $time = null)
...
      */
-    public function generateSecret()
+    public function generateSecret(): string
     {
-        $secret = random_bytes($this->secretLength);
-
-        $base32 = new FixedBitNotation(5, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567', true, true);
-
-        return $base32->encode($secret);
+        return (new FixedBitNotation(5, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567', true, true))->encode(random_bytes($this->secretLength));
```
Comment (OskarStark) [inline src/GoogleAuthenticator.php]: can you please linebreak this

File: tests/GoogleAuthenticatorTest.php [hunk 3]
```diff
@@ -33,10 +33,51 @@ public function testGenerateSecret()
         );
     }
 
-    public function testGetCode()
+    public function testCheckCode()
```
Comment (OskarStark) [inline tests/GoogleAuthenticatorTest.php]: can we use a dataProvider here?

**Comment by OskarStark** [comment]
Close/Reopen for new StyleCI trigger...

File: tests/GoogleAuthenticatorTest.php [hunk 4]
```diff
@@ -33,52 +33,28 @@ public function testGenerateSecret()
...
-        $this->assertFalse(
-            $authenticator->checkCode('3DHTQX4GCRKHGS55CJ', $authenticator->getCode('3DHTQX4GCRKHGS55CJ', false))
-        );
+        return [
+            [false, '2012-03-17 22:16:29'],
+            [true, '2012-03-17 22:16:30'],
+            [true, '2012-03-17 22:17:00'],
+            [true, '2012-03-17 22:17:30'],
+            [false, '2012-03-17 22:18:00'],
+            [false, 'this date cannot be resolved and results into false'],
```
Comment (greg0ire) [inline tests/GoogleAuthenticatorTest.php]: Please use keys in this array to describe each case as you did with the method names

Comment (iltar) [inline tests/GoogleAuthenticatorTest.php]: I can add a generic description to the test (or data provider) itself, which I think is more useful

Comment (greg0ire) [inline tests/GoogleAuthenticatorTest.php]: Do as you think is best

File: src/GoogleAuthenticator.php [hunk 5]
```diff
@@ -113,26 +117,28 @@ public function getCode($secret, $time = null)
...
      */
-    public function generateSecret()
+    public function generateSecret(): string
     {
-        $secret = random_bytes($this->secretLength);
-
-        $base32 = new FixedBitNotation(5, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567', true, true);
-
-        return $base32->encode($secret);
+        return (new FixedBitNotation(5, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567', true, true))
```
Comment (jordisala1991) [inline src/GoogleAuthenticator.php]: Isn't this line a bit too much of: "not creating useless variables"?

File: src/FixedBitNotation.php [hunk 6]
```diff
@@ -193,7 +193,7 @@ public function encode($rawString)
      * @param bool   $strict        Returns null if $encodedString contains
      *                              an undecodable character
      *
-     * @return string|null
+     * @return string
      */
     public function decode($encodedString, $caseSensitive = true, $strict = false)
```
Comment (jordisala1991) [inline src/FixedBitNotation.php]: This should return string right?

Comment (iltar) [inline src/FixedBitNotation.php]: I could add a return statement here yes

File: src/GoogleAuthenticator.php (hunk 5 above)
Comment (iltar) [inline src/GoogleAuthenticator.php]: I don't see how there can be too much of that, saves a variable which has only 1 purpose, to call 1 method and then never use the object ever again

Comment (jordisala1991) [inline src/GoogleAuthenticator.php]: Well, for me its a bit confusing, but I am not against it...

Just pointing out :)

Comment (greg0ire) [inline src/GoogleAuthenticator.php]: Couldn't agree more with @iltar : it just puts distance between usage and definition.

**Comment by greg0ire** [comment]
Thanks a lot @iltar !

**Pull request merged**